### ISSUE TITLE
Unexpected nullable warning on expression used in tuple assignment

### ISSUE THREAD

**Issue by lorcanmooney** [dotnet/roslyn#77245]
**Version Used**: 
17.13.0

**Steps to Reproduce**:

```csharp
#nullable enable

Int32 i;
C? c = null;

if (c is not null)
{
    c = c.NoTuple();    // No warning (as expected)
}

if (c is not null)
{
    (c, i) = c.Tuple(); // Unexpected CS8602 warning on 'c'
}

class C
{
    public C? NoTuple() => throw new NotImplementedException();

    public (C?, Int32) Tuple() => throw new NotImplementedException();
}
```

**Diagnostic Id**:
CS8602: Dereference of a possibly null reference.

**Expected Behavior**:
No warnings.

**Actual Behavior**:
A warning on a variable known to be non-null.

**Notes**:
Swapping the order of the tuple elements doesn't produce the warning.

**Comment by boxofyellow** [comment]
I bumped into this too.  And I was checking to see if it had already been reported.  I was not expecting to find that it had been reported less that 8 hours ago 😄

**Comment by RikkiGibson** [comment]
Related: #36946

> This gets even worse/more strange with deconstruction as Roslyn thinks that assignment of a deconstruction result happens before the deconstruction itself.
